okhttp-tests/src/test/java/okhttp3/HeadersTest.java

```diff
@@ -434,43 +482,254 @@
+    assertEquals(1, challenges.size());
+    assertEquals("digest", challenges.get(0).scheme());
+    assertEquals("myrealm", challenges.get(0).realm());
+    expectedAuthParams.clear();
+    expectedAuthParams.put("realm", "myrealm");
+    assertEquals(expectedAuthParams, challenges.get(0).authParams());
 
     // Scheme only.
     headers = new Headers.Builder()
         .add("WWW-Authenticate", "Digest").build();
     challenges = HttpHeaders.parseChallenges(headers, "WWW-Authenticate");
-    assertEquals(0, challenges.size());
+    assertEquals(1, challenges.size());
+    assertEquals("digest", challenges.get(0).scheme());
+    assertNull(challenges.get(0).realm());
+    expectedAuthParams.clear();
+    assertEquals(expectedAuthParams, challenges.get(0).authParams());
   }
 
   @Test public void basicChallenge() {
     Headers headers = new Headers.Builder()
         .add("WWW-Authenticate: Basic realm=\"protected area\"")
         .build();
-    assertEquals(Arrays.asList(new Challenge("Basic", "protected area")),
+    assertEquals(singletonList(new Challenge("Basic", singletonMap("realm", "protected area"))),
         HttpHeaders.parseChallenges(headers, "WWW-Authenticate"));
   }
 
   @Test public void basicChallengeWithCharset() {
     Headers headers = new Headers.Builder()
         .add("WWW-Authenticate: Basic realm=\"protected area\", charset=\"UTF-8\"")
         .build();
-    assertEquals(Arrays.asList(new Challenge("Basic", "protected area").withCharset(Util.UTF_8)),
+    Map<String, String> expectedAuthParams = new HashMap<>();
+    expectedAuthParams.put("realm", "protected area");
+    expectedAuthParams.put("charset", "UTF-8");
+    assertEquals(singletonList(new Challenge("Basic", expectedAuthParams)),
         HttpHeaders.parseChallenges(headers, "WWW-Authenticate"));
   }
 
   @Test public void basicChallengeWithUnexpectedCharset() {
     Headers headers = new Headers.Builder()
         .add("WWW-Authenticate: Basic realm=\"protected area\", charset=\"US-ASCII\"")
         .build();
-    assertEquals(Collections.emptyList(), HttpHeaders.parseChallenges(headers, "WWW-Authenticate"));
+
+    Map<String, String> expectedAuthParams = new HashMap<>();
+    expectedAuthParams.put("realm", "protected area");
+    expectedAuthParams.put("charset", "US-ASCII");
+    assertEquals(singletonList(new Challenge("Basic", expectedAuthParams)),
+        HttpHeaders.parseChallenges(headers, "WWW-Authenticate"));
+  }
+
+  @Test public void separatorsBeforeFirstChallenge() {
+    Headers headers = new Headers.Builder()
+        .add("WWW-Authenticate", " ,  , Basic realm=myrealm")
+        .build();
+    assertEquals(singletonList(new Challenge("Basic", singletonMap("realm", "myrealm"))),
+        HttpHeaders.parseChallenges(headers, "WWW-Authenticate"));
+  }
+
+  @Test public void spacesAroundKeyValueSeparator() {
+    Headers headers = new Headers.Builder()
+            .add("WWW-Authenticate", "Basic realm = \"myrealm\"")
```

Comment:
nit: https://github.com/square/java-code-styles/

Reply:
You mean the indentation I guess? My bad, fixed.